libavfilter/sf_text2graphicsub.c

```diff
@@ -0,0 +1,492 @@
+{
+    PalettizeContext *context = *palettizecontext;
+    av_freep(&context->codebook);
+    av_freep(&context->codeword);
+    av_freep(&context->codeword_closest_codebook_idxs);
+    avpriv_elbg_free(&context->elbg);
+    av_free(context);
+    *palettizecontext = NULL;
+}
+
+// The following functions and structures are copied from vf_subtitles.c (w/ some minor changes)
+
+/* libass supports a log level ranging from 0 to 7 */
+static const int ass_libavfilter_log_level_map[] = {
+        [0] = AV_LOG_FATAL,     /* MSGL_FATAL */
+        [1] = AV_LOG_ERROR,     /* MSGL_ERR */
+        [2] = AV_LOG_WARNING,   /* MSGL_WARN */
+        [3] = AV_LOG_WARNING,   /* <undefined> */
+        [4] = AV_LOG_INFO,      /* MSGL_INFO */
+        [5] = AV_LOG_INFO,      /* <undefined> */
+        [6] = AV_LOG_VERBOSE,   /* MSGL_V */
+        [7] = AV_LOG_DEBUG,     /* MSGL_DBG2 */
+};
+
+static void ass_log(int ass_level, const char *fmt, va_list args, void *ctx)
+{
+    const int ass_level_clip = av_clip(ass_level, 0,
+        FF_ARRAY_ELEMS(ass_libavfilter_log_level_map) - 1);
+    const int level = ass_libavfilter_log_level_map[ass_level_clip];
+    av_vlog(NULL, level, fmt, args);
+    av_log(NULL, level, "\n");
+}
+
+static const char * const font_mimetypes[] = {
+        "font/ttf",
+        "font/otf",
+        "font/sfnt",
+        "font/woff",
+        "font/woff2",
+        "application/font-sfnt",
+        "application/font-woff",
+        "application/x-truetype-font",
+        "application/vnd.ms-opentype",
+        "application/x-font-ttf",
+        NULL
+};
+
+static int stream_is_font(AVStream* st)
+{
+    const AVDictionaryEntry *tag = NULL;
+    int n;
+
+    if (st->codecpar->codec_type != AVMEDIA_TYPE_ATTACHMENT)
+        return 0;
+
+    tag = av_dict_get(st->metadata, "mimetype", NULL, AV_DICT_MATCH_CASE);
+
+    if (tag) {
+        for (n = 0; font_mimetypes[n]; n++) {
+            if (av_strcasecmp(font_mimetypes[n], tag->value) == 0)
+                return 1;
+        }
+    }
+    return 0;
+}
+
+// Filter structures
+
+#define OFFSET(x) offsetof(ASS_Context, x)
+#define FLAGS (AV_OPT_FLAG_SUBTITLE_PARAM|AV_OPT_FLAG_FILTERING_PARAM)
+
+static const AVOption text2graphicsub_options[] = {
+        { "s",          "output size",                          OFFSET(size),       AV_OPT_TYPE_IMAGE_SIZE,                         .default_val.str = NULL, .flags = FLAGS },
+        { "size",       "output size",                          OFFSET(size),       AV_OPT_TYPE_IMAGE_SIZE,                         .default_val.str = NULL, .flags = FLAGS },
+        { "n",          "number of output colors",              OFFSET(num_colors), AV_OPT_TYPE_INT,        .min = 2, .max = 256,   .default_val.i64 = 256,  .flags = FLAGS },
+        { "num_colors", "number of output colors",              OFFSET(num_colors), AV_OPT_TYPE_INT,        .min = 2, .max = 256,   .default_val.i64 = 256,  .flags = FLAGS },
```

Comment:
After testing, I come to the conclusion that 256 is not a useful default.

1. It creates a huge increase in processing time 
2. Like I told you already: "Our" quantization is **by far** superior to what the dvdsub encoder does

Just compare the transcoding time and output quality when you set num_colors=4 (when encoding dvd_sub)

Reply:
Yeah, I forgot to change that.